### PR TITLE
[dhctl] Break circle and output error in log on check dependencies if get first error

### PR DESCRIPTION
**Pull request by d2285**
## Description
Break circle goroutines if get error on check dependencies, examp. not exist bin file or tcp interrupt connection

## Why do we need it, and what problem does it solve?
this allow to break 30 attempts check dependencies on error and log it to output

## What is the expected result?
check dependencies finish on first error and output log in terminal

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: dhctl
type: fix
summary: Break circle and output error in log on check dependencies if get first error
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 1]
```diff
@@ -576,29 +581,35 @@ func CheckDHCTLDependencies(nodeInteface node.Interface) error {
...
 			wg.Wait()
 			log.DebugLn("Close result chan")
 			close(resultsChan)
 		}()
 
 		for res := range resultsChan {
+
 			if res.err != nil {
-				return fmt.Errorf("Dependency '%s' check failed. Last error %v", res.name, res.err)
+				if strings.Contains(res.err.Error(), "All dependency checks was exceed") {
```
Comment (mvasl) [inline dhctl/pkg/operations/bootstrap/steps.go:599]: Use [errors.Is](https://pkg.go.dev/errors#Is), Luke

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 2]
```diff
@@ -540,49 +540,79 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
+
 	type checkResult struct {
 		name string
 		err  error
 	}
 
-	checkDependency := func(dep string) error {
+	checkDependency := func(dep string, resultsChan chan checkResult) error {
+
+		//breakPredicate := func(err error) bool {
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:551]: remove dead code, please

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 3]
```diff
@@ -540,49 +540,79 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
 			}
-
 			return nil
 		})
 	}
 
 	return log.Process("bootstrap", "Check DHCTL Dependencies", func() error {
-		dependencyArgs := []string{"sudo", "rm", "tar", "mount", "awk", "grep", "cut", "sed", "shopt",
+
+		dependencyCommands := []string{"sudo", "rm", "tar", "mount", "awk", "grep", "cut", "sed", "shopt",
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:587]: better way create slice of slices like 
```go
dependencyCommands := [][]string{
  { "sudo", "rm", "tar", "mount", "awk", },
  { "grep", "cut", "sed", "shopt", "mkdir", },
  { "cp", "join", "cat", "ps", "kill", },
}
```

and range through slices, than calculate indexes in `for` cycle

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 4]
```diff
@@ -540,49 +540,79 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
+						if err != nil {
+							err = errors.Join(exceedDependency, err)
+						}
+						resultsChan <- checkResult{
+							name: dep,
+							err:  err,
+						}
+					}()
+				}
+				time.Sleep(5 * time.Second)
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:615]:
```suggestion
				time.Sleep(2 * time.Second)
```

**Review by name212 — CHANGES_REQUESTED** [review]
two things

File: dhctl/pkg/operations/bootstrap/steps.go (hunk 3 above)
Comment (d2285) [reply dhctl/pkg/operations/bootstrap/steps.go]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/static'...

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 5]
```diff
@@ -540,49 +540,83 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
-				go func() {
-					defer wg.Done()
-					err := checkDependency(dep)
-					resultsChan <- checkResult{
-						name: dep,
-						err:  err,
-					}
-				}()
 
+			for _, args := range dependencyCommands {
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:600]:
```suggestion
			for _, deps := range dependencyCommands {
```

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 6]
```diff
@@ -540,49 +540,83 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
+				for _, args1 := range args {
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:601]:
```suggestion
				for _, dep := range args {
```

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 7]
```diff
@@ -540,49 +540,83 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
+					wg.Add(1)
+					dep := args1
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:603]:
```suggestion
					dep := dep
```

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 8]
```diff
@@ -540,49 +540,83 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
+
 	type checkResult struct {
 		name string
 		err  error
 	}
 
-	checkDependency := func(dep string) error {
-		return retry.NewSilentLoop(fmt.Sprintf("Check dependency %s", dep), 30, 5*time.Second).Run(func() error {
+	checkDependency := func(dep string, resultsChan chan checkResult) error {
+
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:550]:
```suggestion
```

File: dhctl/pkg/operations/bootstrap/steps.go [hunk 9]
```diff
@@ -540,49 +540,83 @@ func setupRPPTunnel(sshClient *ssh.Client) (func(), error) {
...
 			}
-
 			return nil
 		})
 	}
 
 	return log.Process("bootstrap", "Check DHCTL Dependencies", func() error {
-		dependencyArgs := []string{"sudo", "rm", "tar", "mount", "awk", "grep", "cut", "sed", "shopt",
-			"mkdir", "cp", "join", "cat", "ps", "kill"}
+
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/steps.go:586]:
```suggestion
```

File: dhctl/pkg/operations/bootstrap/steps.go (hunk 6 above)
Comment (d2285) [reply dhctl/pkg/operations/bootstrap/steps.go]: done

File: dhctl/pkg/operations/bootstrap/steps.go (hunk 7 above)
Comment (d2285) [reply dhctl/pkg/operations/bootstrap/steps.go]: done

File: dhctl/pkg/operations/bootstrap/steps.go (hunk 8 above)
Comment (d2285) [reply dhctl/pkg/operations/bootstrap/steps.go]: done

File: dhctl/pkg/operations/bootstrap/steps.go (hunk 9 above)
Comment (d2285) [reply dhctl/pkg/operations/bootstrap/steps.go]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @d2285. I've started the workflow for label 'e2e/run/static'...

**Comment by alexey-gavrilov-flant** [comment]
/e2e/abort static;Static;containerd;1.27 10940167087 10940167087-2-con-1-27 /sys/deckhouse-oss/install:pr9679

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @alexey-gavrilov-flant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by alexey-gavrilov-flant** [comment]
/e2e/abort static;Static;containerd;1.27 10940354573 10940354573-1-con-1-27 /sys/deckhouse-oss/install:pr9679

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @alexey-gavrilov-flant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by alexey-gavrilov-flant** [comment]
/e2e/abort static;Static;containerd;1.27 10940167087 10940167087-2-con-1-27 /sys/deckhouse-oss/install:pr9679

**Pull request merged**